src/frontend/src/binder/expr/binary_op.rs

```diff
@@ -95,8 +95,48 @@ impl Binder {
             BinaryOperator::HashArrow => ExprType::JsonbExtractPath,
             BinaryOperator::HashLongArrow => ExprType::JsonbExtractPathText,
             BinaryOperator::Prefix => ExprType::StartsWith,
-            BinaryOperator::Contains => ExprType::JsonbContains,
-            BinaryOperator::Contained => ExprType::JsonbContained,
+            BinaryOperator::Contains => {
+                let left_type = (!bound_left.is_untyped()).then(|| bound_left.return_type());
+                let right_type = (!bound_right.is_untyped()).then(|| bound_right.return_type());
+                match (left_type, right_type) {
+                    (Some(DataType::List { .. }), Some(DataType::List { .. }))
+                    | (Some(DataType::List { .. }), None)
+                    | (None, Some(DataType::List { .. })) => ExprType::ArrayContains,
+                    (Some(DataType::Jsonb), Some(DataType::Jsonb))
+                    | (Some(DataType::Jsonb), None)
+                    | (None, Some(DataType::Jsonb)) => ExprType::JsonbContains,
+                    (None, None) => ExprType::ArrayContained,
```

Comment:
Considering that `array_contains` and `jsonb_contains` share the same `@>` operator, should we merge their `ExprType` into one `Contains`?

Reply:
Good point. To be honest I do not have a clear rule in mind on when to reuse `ExprType` and when to introduce a new one. Most of the time I tend to reuse (and made the mistake of `jsonb -> unknown-as-text` vs `jsonb #> unknown-as-text[]`), except for `||`. It is clear this PR followed the approach of `||`.

`||` is very special. Before we consider the easy case of `bytea || bytea -> bytea` or `jsonb || jsonb -> jsonb`, there is already `text || anynonarray → text` and `anycompatible || anycompatiblearray → anycompatiblearray`, and `text` has an [advantage](https://www.postgresql.org/docs/16/typeconv-oper.html#:~:text=This%20bias%20towards%20string%20is%20appropriate%20since%20an%20unknown%2Dtype%20literal%20looks%20like%20a%20string.) during the process. We feel it is cleaner to keep the overload resolution logic (especially involving different kinds of **any**) in one place, so that all other places can just use concrete `ExprType` (`TextConcatOp`, `ArrayCat`, `ArrayAppend`, `ArrayPrepend`) without worrying about it again.

Other than the special case of text or array, it seems totally fine to reuse the same `ExprType` for `bytea || bytea -> bytea` (unsupported yet) or `jsonb || jsonb -> jsonb`. Coming back to `@>`, it seems okay to merge right now but I am not sure whether keeping them separated could make support of arrays with different dimensions easier. So I prefer not to merge here if that brings little benefit to us.

For reference, PostgreSQL has no concept of `ExprType` at all, and dispatches to different implementations very early. It even has `booleq`, `int2eq`, `int4eq`, `int24eq`, `int28eq`, ... As of now we prefer not to follow this practice and relies on some short hacks (#2680) when compatibility is required.